### ISSUE TITLE
[Bug]: Refine channel (re)assignment logic

### ISSUE THREAD

**Issue by soothing-rain** [milvus-io/milvus#17270]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: latest master
- Deployment mode(standalone or cluster): cluster
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

It is discovered in #15966 that when a RootCoord pod is killed in a Milvus cluster, flush will hang.

This is what happened:
1. RootCoord pod got killed and crashed DataNode-A (There is a bug in code).
2. DataNode was restarted by k8s immediately, with a different node ID, say DataNode-B.
3. DataCoord saw DataNode-B but didn't yet know that DataNode-A was offline and considered both DataNode-A and DataNode-B are alive and reassigned some of DataNode-A's channels to DataNode-B.
4. DataNode-B started to watch the newly assigned channel, which failed because RootCoord is offline.
5. All DataNode-B channels are re-assigned back to DataNode-A.
6. DataCoord finally found that DataNode-A is offline.
7. Reassigning DataNode-B channels to DataNode-A failed, and since there were no other DataNode to assign channels to, these channels were reassigned to the buffer channel.
8. Since channels in buffer only gets reassigned when new DataNodes are discovered, these channels then got stuck in buffer channels.
9. DataCoord never heard from these channels since. All flushes on segments in these channels will be no-op, causing flush to hang.

Basically there are at least 4 things that needs to be fixed here.
1. RootCoord pod kill panics DataNode. (a bug in allocIDBatch() method)
2. Buffered channels cannot be re-allocated unless new DataNode comes in.
3. When a DataNode failed to watch channels, it reassigns these channels to another DataNode, but there's no way to re-balance these channel afterwards.
4. There are duplicate channels in a ChannelOpSet.

Related logs:
https://github.com/milvus-io/milvus/suites/6675992625/artifacts/252967718

### Expected Behavior

All mentioned issues should be fixed.

**Comment by XuanYang-cn** [comment]
/assign
/unassign @yanliang567 

related to #16114

**Comment by XuanYang-cn** [comment]
- [ ] RootCoord crashing should not panic DataNode, DataNode should just report `fail to start flowgraph`

**Comment by XuanYang-cn** [comment]
- [ ] Reassignments should reassign a channel back to the old DataNode if no other nodes available

**Comment by soothing-rain** [comment]
> 

DataNode-B failed to watch the new channels because there's a piece of code that looks for the v-channel start position to recover from. Collection start position will be used if no segment position is found:
https://github.com/milvus-io/milvus/blob/9ce9c6ea68cb248748c5cfd527b27aa7a8dec97f/internal/datacoord/handler.go#L91-L97

full stack trace:
```
/go/src/github.com/milvus-io/milvus/internal/distributed/rootcoord/client/client.go:204 github.com/milvus-io/milvus/internal/distributed/rootcoord/client.(*Client).DescribeCollection

/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:823 github.com/milvus-io/milvus/internal/datacoord.(*Server).loadCollectionFromRootCoord

/go/src/github.com/milvus-io/milvus/internal/datacoord/handler.go:144 github.com/milvus-io/milvus/internal/datacoord.(*ServerHandler).GetCollection

/go/src/github.com/milvus-io/milvus/internal/datacoord/handler.go:93 github.com/milvus-io/milvus/internal/datacoord.(*ServerHandler).GetVChanPositions

/go/src/github.com/milvus-io/milvus/internal/datacoord/channel_manager.go:419 github.com/milvus-io/milvus/internal/datacoord.(*ChannelManager).fillChannelWatchInfoWithState

/go/src/github.com/milvus-io/milvus/internal/datacoord/channel_manager.go:553 github.com/milvus-io/milvus/internal/datacoord.(*ChannelManager).updateWithTimer

/go/src/github.com/milvus-io/milvus/internal/datacoord/channel_manager.go:391 github.com/milvus-io/milvus/internal/datacoord.(*ChannelManager).Watch

/go/src/github.com/milvus-io/milvus/internal/datacoord/services.go:953 github.com/milvus-io/milvus/internal/datacoord.(*Server).WatchChannels
```

Regarding buffered channels, @XuanYang-cn will be refining this channel assigning part together with https://github.com/milvus-io/milvus/issues/16114

**Comment by XuanYang-cn** [comment]
http://10.100.32.144:8080/blue/organizations/jenkins/chaos_test_repeat/detail/chaos_test_repeat/20/pipeline/
http://10.100.32.144:8080/blue/organizations/jenkins/chaos_test_repeat/detail/chaos_test_repeat/19/pipeline

I've started 10 rootcoord pod-kill chaos tests with #17294 and #17293, they all passed

So after these two PRs are both merged, @soothing-rain please help to verify if this issue is able to close.

**Comment by xiaofan-luan** [comment]
close it since it should be fixed, wait to be verified in 15966
